For Pulumi Python SDK add to ValueError and AssertionError messages information about resource_name and argumentname

### What happened?

Debugging in Python SDK which resource is failing could be a nightmare because Traceback is leading to pulumi-language-python-exec and not to user program code. Error messages like 'ValueError('unexpected input of type Dummy')' are not very useful.
### 

### Expected Behavior

I would like to get an error message containing information about the resource causing the issue and witch it's property is causing an issue in the format like:

ValueError: During processing resource: 'pda_ec1_devops_test_1_flowlog_BucketOwnershipControls_b' of type <pulumi_aws.s3.bucket_ownership_controls.BucketOwnershipControls object at 0x7fd5323d6cd0> ValueError has risen: 'ValueError("During processing input field called: 'bucket' of type <config_to_resources.Dummy object at 0x7fd5327ca5b0> ValueError has risen: 'ValueError('unexpected input of type Dummy')'")'

### Steps to reproduce

try to create resources like this:
`
class Dummy:
    def __init__(self):
        pass


dd = Dummy()

pulumi_aws.s3.BucketOwnershipControls(
        resource_name="pda_ec1_devops_test_1_flowlog_BucketOwnershipControls_c",
        args=pulumi_aws.s3.BucketOwnershipControlsArgs(
            bucket=dd,
            rule=pulumi_aws.s3.BucketOwnershipControlsRuleArgs(
                object_ownership="BucketOwnerEnforced"
            )
        )
    )
# or this
pulumi_aws.s3.Bucket(
    resource_name="pda_ec1_devops_test_1_flowlog_s3",
    args=pulumi_aws.s3.BucketArgs(
        bucket="some-bucket",
        lifecycle_rules=[
            pulumi_aws.s3.BucketLifecycleRuleArgs(
                enabled=True,
                expiration=pulumi_aws.s3.BucketLifecycleRuleExpirationArgs(
                    days=365
                ),
                id="CleanupRule"
            )
        ],
        tags=[{ #issue here list instead of dict when you mix pulumi_aws with pulumi_aws_native 
            "Name": "some-bucket",
            "SomeTag": "xyz"
        }],
    )
)

`

### Output of `pulumi about`

CLI          
Version      3.74.0
Go Version   go1.20.5
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.41.0
aws-native  0.67.0
python      unknown

Host     
OS       ubuntu
Version  20.04
Arch     x86_64

This project is written in python: executable='/home/adgo/.pyenv/shims/python3' version='3.8.10
'

Current Stack: name

Found no resources associated with name

Found no pending operations associated with name

Backend        
Name         ***
URL            s3
User           ***
Organizations  

Dependencies:
NAME                      VERSION
boto3                     1.26.165
mypy                      1.4.1
pip                       23.1.2
pkg_resources             0.0.0
pulumi-aws                5.41.0
pulumi-aws-native         0.67.0
pydantic                  2.0.2
service-directory-loader  1.0.4
setuptools                44.0.0


### Additional context

Proposed improvements
1)
https://github.com/pulumi/pulumi/blob/86a240362955ecf64752841d65853194ad8a9a44/sdk/python/lib/pulumi/runtime/rpc.py#L209
wrap with try ... except block like : 

        try:
            result = await serialize_property(
                v, deps, input_transformer, get_type(k), keep_output_values
            )
        except ValueError as e: 
            raise ValueError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
        except AssertionErroras e: 
            raise AssertionError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)

2)
https://github.com/pulumi/pulumi/blob/86a240362955ecf64752841d65853194ad8a9a44/sdk/python/lib/pulumi/runtime/resource.py#L872
wrap with try ... except block like : 
            try:
                resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
            except ValueError  as e: 
                raise ValueError(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
           except AssertionError: as e:
                raise AssertionError:(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already). 

see additional content with proposed improvements

--- a/changelog/pending/20241018--sdk-python--fixes-13516-fix-valueerror-and-assertionerror-to-provide-better-debugging-information.yaml
+++ b/changelog/pending/20241018--sdk-python--fixes-13516-fix-valueerror-and-assertionerror-to-provide-better-debugging-information.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: sdk/python
+  description: "Fixes #13516 - fix ValueError and AssertionError to provide better debugging information"

--- a/changelog/pending/20241018--sdk-python--fixes-13516-fix-valueerror-and-assertionerror-to-provide-better-debugging-information.yaml
+++ b/changelog/pending/20241018--sdk-python--fixes-13516-fix-valueerror-and-assertionerror-to-provide-better-debugging-information.yaml
@@ -1,4 +1,4 @@
 changes:
 - type: fix
   scope: sdk/python
-  description: "Fixes #13516 - fix ValueError and AssertionError to provide better debugging information"
+  description: "Add additional debug information to `ValueError` and `AssertionError`"

--- a/sdk/python/lib/pulumi/runtime/resource.py
+++ b/sdk/python/lib/pulumi/runtime/resource.py
@@ -935,9 +935,9 @@ def register_resource(
             try:
                 resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
             except ValueError as e:
-                raise ValueError(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+                raise ValueError(f"During processing resource: {repr(name)} of type {repr(ty)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
             except AssertionError as e:
-                raise AssertionError(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+                raise AssertionError(f"During processing resource: {repr(name)} of type {repr(ty)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
             log.debug(f"resource registration prepared: ty={ty}, name={name}")
 
             callbacks: List[callback_pb2.Callback] = []

--- a/sdk/python/lib/pulumi/runtime/resource.py
+++ b/sdk/python/lib/pulumi/runtime/resource.py
@@ -932,7 +932,12 @@ def register_resource(
                     opts.plugin_download_url = None
                     opts.version = None
 
-            resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
+            try:
+                resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
+            except ValueError as e:
+                raise ValueError(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+            except AssertionError as e:
+                raise AssertionError(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
             log.debug(f"resource registration prepared: ty={ty}, name={name}")
 
             callbacks: List[callback_pb2.Callback] = []

--- a/sdk/python/lib/pulumi/runtime/resource.py
+++ b/sdk/python/lib/pulumi/runtime/resource.py
@@ -940,12 +940,12 @@ def register_resource(
                 raise ValueError(
                     f"While processing resource: {repr(name)}, type: {repr(ty)}\n"
                     + f"ValueError has risen: {e}"
-                )
+                ) from e
             except AssertionError as e:
                 raise AssertionError(
                     f"While processing resource: {repr(name)}, type {repr(ty)}\n"
                     + f"AssertionError has risen: {e}"
-                )
+                ) from e
             log.debug(f"resource registration prepared: ty={ty}, name={name}")
 
             callbacks: List[callback_pb2.Callback] = []

--- a/sdk/python/lib/pulumi/runtime/resource.py
+++ b/sdk/python/lib/pulumi/runtime/resource.py
@@ -935,9 +935,15 @@ def register_resource(
             try:
                 resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
             except ValueError as e:
-                raise ValueError(f"During processing resource: {repr(name)} of type {repr(ty)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+                raise ValueError(
+                    f"While processing resource: {repr(name)}, type: {repr(ty)}\n"
+                    + f"ValueError has risen: {e}"
+                )
             except AssertionError as e:
-                raise AssertionError(f"During processing resource: {repr(name)} of type {repr(ty)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+                raise AssertionError(
+                    f"While processing resource: {repr(name)}, type {repr(ty)}\n"
+                    + f"AssertionError has risen: {e}"
+                )
             log.debug(f"resource registration prepared: ty={ty}, name={name}")
 
             callbacks: List[callback_pb2.Callback] = []

--- a/sdk/python/lib/pulumi/runtime/resource.py
+++ b/sdk/python/lib/pulumi/runtime/resource.py
@@ -933,7 +933,9 @@ def register_resource(
                     opts.version = None
 
             try:
-                resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
+                resolver = await prepare_resource(
+                    res, ty, custom, remote, props, opts, typ
+                )
             except ValueError as e:
                 raise ValueError(
                     f"While processing resource: {repr(name)}, type: {repr(ty)}\n"

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -204,10 +204,14 @@ async def serialize_properties(
     if typ is not None:
         py_name_to_pulumi_name = _types.input_type_py_to_pulumi_names(typ)
         types = _types.input_type_types(typ)
+
         # pylint: disable=C3001
-        def translate(k): return py_name_to_pulumi_name.get(k) or k
+        def translate(k):
+            return py_name_to_pulumi_name.get(k) or k
+
         # pylint: disable=C3001
-        def get_type(k): return types.get(translate(k))  # type: ignore
+        def get_type(k):
+            return types.get(translate(k))  # type: ignore
 
     struct = struct_pb2.Struct()
     # We're deliberately not using `inputs.items()` here in case inputs is a subclass of `dict` that redefines items.
@@ -630,8 +634,11 @@ async def serialize_property(
                 py_name_to_pulumi_name = _types.input_type_py_to_pulumi_names(typ)
                 # pylint: disable=C3001
                 types = _types.input_type_types(typ)
+
                 # pylint: disable=C3001
-                def translate(k): return py_name_to_pulumi_name.get(k) or k
+                def translate(k):
+                    return py_name_to_pulumi_name.get(k) or k
+
                 get_type = types.get
             else:
                 # Otherwise, don't do any translation of user-defined dict keys.
@@ -640,7 +647,9 @@ async def serialize_property(
                     args = get_args(typ)
                     if len(args) == 2 and args[0] is str:
                         # pylint: disable=C3001
-                        def get_type(k): return args[1]
+                        def get_type(k):
+                            return args[1]
+
                         translate = None
                 else:
                     translate = None
@@ -1206,12 +1215,16 @@ def translate_output_properties(
                 args = get_args(typ)
                 if len(args) == 2 and args[0] is str:
                     # pylint: disable=C3001
-                    def get_type(k): return args[1]
+                    def get_type(k):
+                        return args[1]
+
                     # If transform_using_type_metadata is True, don't translate its keys because
                     # it is intended to be a user-defined dict.
                     if transform_using_type_metadata:
                         # pylint: disable=C3001
-                        def translate(k): return k
+                        def translate(k):
+                            return k
+
             elif return_none_on_dict_type_mismatch:
                 return None
             else:
@@ -1350,10 +1363,14 @@ def resolve_outputs(
     )
     if transform_using_type_metadata:
         pulumi_to_py_names = _types.resource_pulumi_to_py_names(resource_cls)
+
         # pylint: disable=C3001
-        def translate(prop): return pulumi_to_py_names.get(prop) or prop
+        def translate(prop):
+            return pulumi_to_py_names.get(prop) or prop
+
         # pylint: disable=C3001
-        def translate_to_pass(prop): return prop
+        def translate_to_pass(prop):
+            return prop
 
     for key, value in deserialize_properties(outputs, keep_unknowns).items():
         # Outputs coming from the provider are NOT translated. Do so here.

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -214,15 +214,20 @@ async def serialize_properties(
     for k in inputs:
         v = inputs[k]
         deps: List["Resource"] = []
-        result = await serialize_property(
-            v,
-            deps,
-            k,
-            resource_obj,
-            input_transformer,
-            get_type(k),
-            keep_output_values,
-        )
+        try:
+            result = await serialize_property(
+                v,
+                deps,
+                k,
+                resource_obj,
+                input_transformer,
+                get_type(k),
+                keep_output_values,
+            )
+        except ValueError as e: 
+            raise ValueError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+        except AssertionError as e: 
+            raise AssertionError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
         # We treat properties that serialize to None as if they don't exist.
         if result is not None:
             # While serializing to a pb struct, we must "translate" all key names to be what the

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -205,9 +205,9 @@ async def serialize_properties(
         py_name_to_pulumi_name = _types.input_type_py_to_pulumi_names(typ)
         types = _types.input_type_types(typ)
         # pylint: disable=C3001
-        translate = lambda k: py_name_to_pulumi_name.get(k) or k
+        def translate(k): return py_name_to_pulumi_name.get(k) or k
         # pylint: disable=C3001
-        get_type = lambda k: types.get(translate(k))  # type: ignore
+        def get_type(k): return types.get(translate(k))  # type: ignore
 
     struct = struct_pb2.Struct()
     # We're deliberately not using `inputs.items()` here in case inputs is a subclass of `dict` that redefines items.
@@ -225,9 +225,15 @@ async def serialize_properties(
                 keep_output_values,
             )
         except ValueError as e:
-            raise ValueError(f"During processing input field called: {repr(k)} of type {type(k)} with value {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+            raise ValueError(
+                f"While processing input: {repr(k)}, type {type(k)}, value {repr(v)}\n"
+                + f"ValueError has risen: {e}"
+            )
         except AssertionError as e:
-            raise AssertionError(f"During processing input field called: {repr(k)} of type {type(k)} with value {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+            raise AssertionError(
+                f"While processing input: {repr(k)}, type {type(k)}, value {repr(v)}\n"
+                + f"AssertionError has risen: {e}"
+            )
         # We treat properties that serialize to None as if they don't exist.
         if result is not None:
             # While serializing to a pb struct, we must "translate" all key names to be what the
@@ -625,7 +631,7 @@ async def serialize_property(
                 # pylint: disable=C3001
                 types = _types.input_type_types(typ)
                 # pylint: disable=C3001
-                translate = lambda k: py_name_to_pulumi_name.get(k) or k
+                def translate(k): return py_name_to_pulumi_name.get(k) or k
                 get_type = types.get
             else:
                 # Otherwise, don't do any translation of user-defined dict keys.
@@ -634,7 +640,7 @@ async def serialize_property(
                     args = get_args(typ)
                     if len(args) == 2 and args[0] is str:
                         # pylint: disable=C3001
-                        get_type = lambda k: args[1]
+                        def get_type(k): return args[1]
                         translate = None
                 else:
                     translate = None
@@ -1200,12 +1206,12 @@ def translate_output_properties(
                 args = get_args(typ)
                 if len(args) == 2 and args[0] is str:
                     # pylint: disable=C3001
-                    get_type = lambda k: args[1]
+                    def get_type(k): return args[1]
                     # If transform_using_type_metadata is True, don't translate its keys because
                     # it is intended to be a user-defined dict.
                     if transform_using_type_metadata:
                         # pylint: disable=C3001
-                        translate = lambda k: k
+                        def translate(k): return k
             elif return_none_on_dict_type_mismatch:
                 return None
             else:
@@ -1345,9 +1351,9 @@ def resolve_outputs(
     if transform_using_type_metadata:
         pulumi_to_py_names = _types.resource_pulumi_to_py_names(resource_cls)
         # pylint: disable=C3001
-        translate = lambda prop: pulumi_to_py_names.get(prop) or prop
+        def translate(prop): return pulumi_to_py_names.get(prop) or prop
         # pylint: disable=C3001
-        translate_to_pass = lambda prop: prop
+        def translate_to_pass(prop): return prop
 
     for key, value in deserialize_properties(outputs, keep_unknowns).items():
         # Outputs coming from the provider are NOT translated. Do so here.
@@ -1501,7 +1507,8 @@ def register_resource_package(pkg: str, package: ResourcePackage):
         for existing in resource_packages:
             if same_version(existing.version(), package.version()):
                 raise ValueError(
-                    f"Cannot re-register package {pkg}@{package.version()}. Previous registration was {existing}, new registration was {package}."
+                    f"Cannot re-register package {pkg}@{package.version()
+                                                        }. Previous registration was {existing}, new registration was {package}."
                 )
     else:
         resource_packages = []
@@ -1550,7 +1557,8 @@ def register_resource_module(pkg: str, mod: str, module: ResourceModule):
         for existing in resource_modules:
             if same_version(existing.version(), module.version()):
                 raise ValueError(
-                    f"Cannot re-register module {key}@{module.version()}. Previous registration was {existing}, new registration was {module}."
+                    f"Cannot re-register module {key}@{module.version()
+                                                       }. Previous registration was {existing}, new registration was {module}."
                 )
     else:
         resource_modules = []

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -1207,14 +1207,12 @@ def translate_output_properties(
                 args = get_args(typ)
                 if len(args) == 2 and args[0] is str:
                     # pylint: disable=C3001
-                    def get_type(k):
-                        return args[1]
-
+                    get_type = lambda k: args[1]
                     # If transform_using_type_metadata is True, don't translate its keys because
                     # it is intended to be a user-defined dict.
                     if transform_using_type_metadata:
                         # pylint: disable=C3001
-                        get_type = lambda k: args[1]
+                        translate = lambda k: k
 
             elif return_none_on_dict_type_mismatch:
                 return None

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -224,10 +224,10 @@ async def serialize_properties(
                 get_type(k),
                 keep_output_values,
             )
-        except ValueError as e: 
-            raise ValueError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
-        except AssertionError as e: 
-            raise AssertionError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+        except ValueError as e:
+            raise ValueError(f"During processing input field called: {repr(k)} of type {type(k)} with value {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
+        except AssertionError as e:
+            raise AssertionError(f"During processing input field called: {repr(k)} of type {type(k)} with value {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
         # We treat properties that serialize to None as if they don't exist.
         if result is not None:
             # While serializing to a pb struct, we must "translate" all key names to be what the

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -630,10 +630,8 @@ async def serialize_property(
                 py_name_to_pulumi_name = _types.input_type_py_to_pulumi_names(typ)
                 # pylint: disable=C3001
                 types = _types.input_type_types(typ)
-
                 # pylint: disable=C3001
-                def translate(k):
-                    return py_name_to_pulumi_name.get(k) or k
+                translate = lambda k: py_name_to_pulumi_name.get(k) or k
 
                 get_type = types.get
             else:
@@ -643,9 +641,7 @@ async def serialize_property(
                     args = get_args(typ)
                     if len(args) == 2 and args[0] is str:
                         # pylint: disable=C3001
-                        def get_type(k):
-                            return args[1]
-
+                        get_type = lambda k: args[1]
                         translate = None
                 else:
                     translate = None
@@ -1218,8 +1214,7 @@ def translate_output_properties(
                     # it is intended to be a user-defined dict.
                     if transform_using_type_metadata:
                         # pylint: disable=C3001
-                        def translate(k):
-                            return k
+                        get_type = lambda k: args[1]
 
             elif return_none_on_dict_type_mismatch:
                 return None
@@ -1361,12 +1356,9 @@ def resolve_outputs(
         pulumi_to_py_names = _types.resource_pulumi_to_py_names(resource_cls)
 
         # pylint: disable=C3001
-        def translate(prop):
-            return pulumi_to_py_names.get(prop) or prop
-
+        translate = lambda prop: pulumi_to_py_names.get(prop) or prop
         # pylint: disable=C3001
-        def translate_to_pass(prop):
-            return prop
+        translate_to_pass = lambda prop: prop
 
     for key, value in deserialize_properties(outputs, keep_unknowns).items():
         # Outputs coming from the provider are NOT translated. Do so here.
@@ -1520,8 +1512,7 @@ def register_resource_package(pkg: str, package: ResourcePackage):
         for existing in resource_packages:
             if same_version(existing.version(), package.version()):
                 raise ValueError(
-                    f"Cannot re-register package {pkg}@{package.version()
-                                                        }. Previous registration was {existing}, new registration was {package}."
+                    f"Cannot re-register package {pkg}@{package.version()}. Previous registration was {existing}, new registration was {package}."
                 )
     else:
         resource_packages = []
@@ -1570,8 +1561,7 @@ def register_resource_module(pkg: str, mod: str, module: ResourceModule):
         for existing in resource_modules:
             if same_version(existing.version(), module.version()):
                 raise ValueError(
-                    f"Cannot re-register module {key}@{module.version()
-                                                       }. Previous registration was {existing}, new registration was {module}."
+                    f"Cannot re-register module {key}@{module.version()}. Previous registration was {existing}, new registration was {module}."
                 )
     else:
         resource_modules = []

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -232,12 +232,12 @@ async def serialize_properties(
             raise ValueError(
                 f"While processing input: {repr(k)}, type {type(k)}, value {repr(v)}\n"
                 + f"ValueError has risen: {e}"
-            )
+            ) from e
         except AssertionError as e:
             raise AssertionError(
                 f"While processing input: {repr(k)}, type {type(k)}, value {repr(v)}\n"
                 + f"AssertionError has risen: {e}"
-            )
+            ) from e
         # We treat properties that serialize to None as if they don't exist.
         if result is not None:
             # While serializing to a pb struct, we must "translate" all key names to be what the

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -204,14 +204,10 @@ async def serialize_properties(
     if typ is not None:
         py_name_to_pulumi_name = _types.input_type_py_to_pulumi_names(typ)
         types = _types.input_type_types(typ)
-
         # pylint: disable=C3001
-        def translate(k):
-            return py_name_to_pulumi_name.get(k) or k
-
+        translate = lambda k: py_name_to_pulumi_name.get(k) or k
         # pylint: disable=C3001
-        def get_type(k):
-            return types.get(translate(k))  # type: ignore
+        get_type = lambda k: types.get(translate(k))  # type: ignore
 
     struct = struct_pb2.Struct()
     # We're deliberately not using `inputs.items()` here in case inputs is a subclass of `dict` that redefines items.

